### PR TITLE
Wire datum refactor and converts all wire nanoUIs to use TGUI

### LINKED ISSUE #12972: Runtime when deleting something with its wires panel open
```
[2020-02-11T20:02:01] Runtime in camera.dm,16: Cannot read null.panel_open
   proc name: CanUse (/datum/wires/camera/CanUse)
   src: /datum/wires/camera (/datum/wires/camera)
   call stack:
   /datum/wires/camera (/datum/wires/camera): CanUse(Madison Zhang (/mob/living/carbon/human))
   /datum/wires/camera (/datum/wires/camera): CanUseTopic(Madison Zhang (/mob/living/carbon/human), /datum/topic_state/default (/datum/topic_state/default))
   /datum/nanoui (/datum/nanoui): update status(0)
   /datum/nanoui (/datum/nanoui): push data(null, 1)
   /datum/nanoui (/datum/nanoui): set status(-1, 1)
   /datum/nanoui (/datum/nanoui): update status(1)
   /datum/nanoui (/datum/nanoui): process(0)
   Nanoui (/datum/controller/subsystem/nanoui): fire(0)
   Nanoui (/datum/controller/subsystem/nanoui): ignite(0)
   Master (/datum/controller/master): RunQueue()
   Master (/datum/controller/master): Loop()
   Master (/datum/controller/master): StartProcessing(0)
```

Reproduce by opening something's wires panel, then deleting the atom while the panel is open. (Tested on cameras and cyborgs)

Think this one is ancient 😄 

EDIT:

This bit runtimes (specifically line 16):

https://github.com/ParadiseSS13/Paradise/blob/a20ce9b7ced9563f801f2a0b1abcbbb24e50da83/code/datums/wires/camera.dm#L14-L18

I think below line is to blame - it sets `holder = null` before calling `..()`, therefore when the wires calls `CanUse()` it sees holder is null and runtimes
https://github.com/ParadiseSS13/Paradise/blob/a20ce9b7ced9563f801f2a0b1abcbbb24e50da83/code/datums/wires/wires.dm#L49-L51

### PR DESCRIPTION
**Pull request by SteelSlayer**
## What Does This PR Do
Ports TGs way of handling wire datums and ports their wire TGUI interface. Wire functionality remains the same (or it should anyway). The only thing users should notice is the different coloring and of course, that the UI is now TGUI.

Backend things:
TGs wire datums weren't *that* different from what we currently have. So most of the code will look similar. For the most part this is a 1:1 port. There are two things that are different:
- There is quite a lot more code in `tgui_data()`. We need a bunch of extra code to handle colorblindness. 
- The `assemblies` list in the wire datum can hold all types of assemblies (timers, igniters, etc) on TG. But for Paradise, we only use signalers. which is why you'll see signaler arguments instead of all base assemblies. `obj/item/assembly/S` vs
 `obj/item/assembly/signaler/S` for example.

Oh and of course, I tried to add documentation for all procs and vars in the wire datum, which isn't present on TG.

## Why It's Good For The Game
NanoUI is crap, more TGUI is good. Better readability of wire colors. No more eye strain trying to pick out the black and navy wires.
Fixes #12972

## Images of changes
**Disclaimer**: In the screenshots you'll see, I was using an abductor multitool, which shows the wire information like "Electrification" and so on. These are NOT present if you do not have a mutitool that shows that information. The standard yellow ones do not have this effect.

Here is an imgur album with every single new wire UI: https://imgur.com/a/HPb8y2Q

Sample UI for airlocks. This shows the entire new pool of wire colors.
airlock

## Changelog
:cl:
add: The wire UI windows now use TGUI.
tweak: Slightly tweaked the available pool of wire colors for more readability. Navy, black and grey are removed. Magenta, silver and white take their places.
tweak: You can now use a multitool on the particle accelerator console to access the wires. Previously you could only access it by clicking it with an open hand.
/:cl:

### CODE REVIEW

File: code/datums/wires/wires.dm [hunk 1]
```diff
@@ -1,331 +1,456 @@
...
 
 /datum/wires
+	/// TRUE if the wires will be different every time a new wire datum is created.
+	var/randomize = FALSE
+	/// The atom the wires belong too. For example: an airlock.
+	var/atom/holder
+	/// The holder type; used to make sure that the holder is the correct type.
+	var/holder_type
+	/// The display name for the TGUI window. For example, given the var is "APC"...
+	/// When the TGUI window is opened, "wires" will be appended to it's title, and it would become "APC wires".
```
Comment (AffectedArc07) [inline code/datums/wires/wires.dm:10]: I dont know if the double-lined `///` will work with the doc here.

Comment (SteelSlayer) [reply code/datums/wires/wires.dm:10]: Might not. I think I saw this done on TG in a few places, which *seemed* to suggest it was ok but, I dont know for sure.

File: code/datums/wires/apc.dm [hunk 2]
```diff
@@ -29,66 +17,63 @@
...
 
-	switch(index)
-
-		if(APC_WIRE_IDSCAN)
-			A.locked = 0
+	switch(wire)
+		if(WIRE_IDSCAN)
+			A.locked = FALSE
 
 			spawn(300)
```
Comment (AffectedArc07) [inline code/datums/wires/apc.dm:34]: Same here

File: code/datums/wires/apc.dm [hunk 3]
```diff
@@ -29,66 +17,63 @@
...
 					A.updateDialog()
 
-		if(APC_WIRE_MAIN_POWER1, APC_WIRE_MAIN_POWER2)
-			if(A.shorted == 0)
-				A.shorted = 1
+		if(WIRE_MAIN_POWER1, WIRE_MAIN_POWER2)
+			if(!A.shorted)
+				A.shorted = TRUE
 
 				spawn(1200)
```
Comment (AffectedArc07) [inline code/datums/wires/apc.dm:43]: And here

File: code/datums/wires/apc.dm [hunk 4]
```diff
@@ -29,66 +17,63 @@
...
 						A.updateDialog()
 
-		if(APC_WIRE_AI_CONTROL)
-			if(A.aidisabled == 0)
-				A.aidisabled = 1
+		if(WIRE_AI_CONTROL)
+			if(!A.aidisabled)
+				A.aidisabled = TRUE
 
 				spawn(10)
```
Comment (AffectedArc07) [inline code/datums/wires/apc.dm:52]: And here

File: code/datums/wires/autolathe.dm [hunk 5]
```diff
@@ -24,51 +16,43 @@
...
+	if(is_cut(wire))
 		return
 	var/obj/machinery/autolathe/A = holder
-	switch(index)
-		if(AUTOLATHE_HACK_WIRE)
+	switch(wire)
+		if(WIRE_AUTOLATHE_HACK)
 			A.adjust_hacked(!A.hacked)
-			updateUIs()
 			spawn(50)
```
Comment (AffectedArc07) [inline code/datums/wires/autolathe.dm:45]: And here

File: code/datums/wires/autolathe.dm [hunk 6]
```diff
@@ -24,51 +16,43 @@
...
-				if(A && !IsIndexCut(index))
+				if(A && !is_cut(wire))
 					A.adjust_hacked(0)
-					updateUIs()
-		if(AUTOLATHE_SHOCK_WIRE)
+		if(WIRE_ELECTRIFY)
 			A.shocked = !A.shocked
-			updateUIs()
 			spawn(50)
```
Comment (AffectedArc07) [inline code/datums/wires/autolathe.dm:50]: And here

File: code/datums/wires/autolathe.dm [hunk 7]
```diff
@@ -24,51 +16,43 @@
...
-				if(A && !IsIndexCut(index))
-					A.shocked = 0
-					updateUIs()
-		if(AUTOLATHE_DISABLE_WIRE)
+				if(A && !is_cut(wire))
+					A.shocked = FALSE
+		if(WIRE_AUTOLATHE_DISABLE)
 			A.disabled = !A.disabled
-			updateUIs()
 			spawn(50)
```
Comment (AffectedArc07) [inline code/datums/wires/autolathe.dm:55]: And here

File: code/datums/wires/nuclearbomb.dm [hunk 8]
```diff
@@ -31,22 +23,19 @@
...
-/datum/wires/nuclearbomb/UpdatePulsed(index)
+/datum/wires/nuclearbomb/on_pulse(wire)
 	var/obj/machinery/nuclearbomb/N = holder
-	switch(index)
-		if(NUCLEARBOMB_WIRE_LIGHT)
+	switch(wire)
+		if(WIRE_BOMB_LIGHT)
 			N.lighthack = !N.lighthack
-			updateUIs()
 			spawn(100)
```
Comment (AffectedArc07) [inline code/datums/wires/nuclearbomb.dm:31]: And here

File: code/datums/wires/nuclearbomb.dm [hunk 9]
```diff
@@ -31,22 +23,19 @@
...
-		if(NUCLEARBOMB_WIRE_TIMING)
+		if(WIRE_BOMB_TIMING)
 			if(N.timing)
 				message_admins("[key_name_admin(usr)] pulsed a nuclear bomb's detonation wire, causing it to explode (<A HREF='?_src_=holder;adminplayerobservecoodjump=1;X=[holder.x];Y=[holder.y];Z=[holder.z]'>JMP</a>)")
 				N.explode()
-		if(NUCLEARBOMB_WIRE_SAFETY)
+		if(WIRE_BOMB_SAFETY)
 			N.safety = !N.safety
-			updateUIs()
 			spawn(100)
```
Comment (AffectedArc07) [inline code/datums/wires/nuclearbomb.dm:39]: And here

File: code/datums/wires/nuclearbomb.dm [hunk 10]
```diff
@@ -58,25 +47,19 @@
...
 			if(N.timing)
 				message_admins("[key_name_admin(usr)] cut a nuclear bomb's timing wire, causing it to explode (<A HREF='?_src_=holder;adminplayerobservecoodjump=1;X=[holder.x];Y=[holder.y];Z=[holder.z]'>JMP</a>)")
 				N.explode()
-		if(NUCLEARBOMB_WIRE_TIMING)
+		if(WIRE_BOMB_TIMING)
 			if(!N.lighthack)
 				if(N.icon_state == "nuclearbomb2")
 					N.icon_state = "nuclearbomb1"
 			N.timing = 0
 			GLOB.bomb_set = 0
```
Comment (AffectedArc07) [inline code/datums/wires/nuclearbomb.dm:63]:
```suggestion
			N.timing = 0
			GLOB.bomb_set = FALSE
```

File: code/datums/wires/particle_accelerator.dm [hunk 11]
```diff
@@ -1,67 +1,52 @@
...
-			if(C.active == !mended)
+	switch(wire)
+		if(WIRE_PARTICLE_POWER)
+			if(C.active == !mend)
 				C.toggle_power()
 
-		if(PARTICLE_STRENGTH_WIRE)
-
+		if(WIRE_PARTICLE_STRENGTH)
 			for(var/i = 1; i < 3; i++)
```
Comment (AffectedArc07) [inline code/datums/wires/particle_accelerator.dm:42]: Would say to use the more efficient loop type here, but then DM would complain about a var being defined but not used. Ah well.

File: code/datums/wires/suitstorage.dm [hunk 12]
```diff
@@ -28,42 +17,44 @@
...
 			A.secure = !A.secure
-		if(SSU_WIRE_SAFETY)
+		if(WIRE_SAFETY)
 			A.safeties = !A.safeties
-		if(SSU_WIRE_SHOCK)
+		if(WIRE_ELECTRIFY)
 			A.shocked = !A.shocked
 			if(A.shocked)
 				A.shock(usr, 100)
 				spawn(50)
```
Comment (AffectedArc07) [inline code/datums/wires/suitstorage.dm:55]: Kill this spawn if possible

File: code/modules/assembly/assembly.dm [hunk 13]
```diff
@@ -1,3 +1,9 @@
+#define WIRE_RECEIVE		(1<<0)	//Allows pulsed(0) to call Activate()
+#define WIRE_PULSE			(1<<1)	//Allows pulse(0) to act on the holder
+#define WIRE_PULSE_SPECIAL	(1<<2)	//Allows pulse(0) to act on the holders special assembly
+#define WIRE_RADIO_RECEIVE	(1<<3)	//Allows pulsed(1) to call Activate()
+#define WIRE_RADIO_PULSE	(1<<4)	//Allows pulse(1) to send a radio message
```
Comment (AffectedArc07) [inline code/modules/assembly/assembly.dm:5]: Mix of `pulse()` and `pulsed()`

**Review by AffectedArc07 — APPROVED** [review]

File: code/datums/wires/alarm.dm [hunk 14]
```diff
@@ -39,74 +26,65 @@
...
-		if(AALARM_WIRE_POWER)
-//			to_chat(world, "Power wire pulsed")
-			if(A.shorted == 0)
-				A.shorted = 1
+		if(WIRE_MAIN_POWER1)
+			if(!A.shorted)
+				A.shorted = TRUE
 				A.update_icon()
 
 			spawn(12000)
```
[the remark this answers is not in the document]
Comment (SteelSlayer) [reply code/datums/wires/alarm.dm:66]: You gonna make me convert all these spawns huh :weary:. Very well.

File: code/datums/wires/particle_accelerator.dm (hunk 11 above)
Comment (SteelSlayer) [reply code/datums/wires/particle_accelerator.dm:42]: I can switch to the other loop with no problems

**Comment by SteelSlayer** [comment]
Should be good to go.

File: code/datums/wires/alarm.dm [hunk 15]
```diff
@@ -39,74 +26,59 @@
...
 
-		if(AALARM_WIRE_POWER)
-//			to_chat(world, "Power wire pulsed")
-			if(A.shorted == 0)
-				A.shorted = 1
+		if(WIRE_MAIN_POWER1)
+			if(!A.shorted)
+				A.shorted = TRUE
 				A.update_icon()
+			addtimer(CALLBACK(A, /obj/machinery/alarm/.proc/unshort_callback), 120 SECONDS)
```
Comment (AffectedArc07) [inline code/datums/wires/alarm.dm:65]: You used `120 SECONDS` here, but the old value was `12000`, which is 12000 seconds. Though saying that, an air alarm being shorted for **20 minutes** seems off, so I prefer this value.

Comment (AffectedArc07) [reply code/datums/wires/alarm.dm:65]: This was spoken about on discord, and is now declared a fix since its the same timer for APCs.

**Pull request merged**